Mathlib/RingTheory/Ideal/Quotient/Index.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.Algebra.GeomSum
+import Mathlib.Data.Finsupp.Fintype
+import Mathlib.GroupTheory.Index
+import Mathlib.LinearAlgebra.DirectSum.Finsupp
+import Mathlib.LinearAlgebra.TensorProduct.Quotient
+import Mathlib.LinearAlgebra.TensorProduct.RightExactness
+import Mathlib.RingTheory.Finiteness.Cardinality
+import Mathlib.RingTheory.Finiteness.TensorProduct
+import Mathlib.RingTheory.Ideal.Quotient.Operations
+
+/-!
+
+# Indices of ideals
+
+## Main result
+- `Submodule.finite_quotient_smul`:
+  Let `N` be a finite index f.g. `R`-submodule, and `I` be a finite index ideal.
+  Then `I • N` also has finite index.
+- `Ideal.index_quotient_pow_le`:
+  If `I` is generated by `k` elements,
+  the index of `I ^ n` is bounded by `#(R ⧸ I) ^ (k⁰ + k¹ + ⋯ + kⁿ⁻¹)`.
+
+-/
+
+variable {R M : Type*} [CommRing R] [AddCommGroup M] [Module R M]
+variable (I : Ideal R) {N : Submodule R M}
+
+open TensorProduct in
+/-- Let `N` be a finite index f.g. `R`-submodule, and `I` be a finite index ideal.
+Then `I • N` also has finite index. -/
+lemma Submodule.finite_quotient_smul [Finite (R ⧸ I)] [Finite (M ⧸ N)] (hN : N.FG) :
+    Finite (M ⧸ I • N) := by
+  suffices (I • N).toAddSubgroup.FiniteIndex by
+    exact (I • N).toAddSubgroup.finite_quotient_of_finiteIndex
+  suffices Nat.card (N ⧸ (I • N).comap N.subtype) ≠ 0 by
+    constructor
+    rw [← AddSubgroup.relindex_mul_index
+      (H := (I • N).toAddSubgroup) (K := N.toAddSubgroup) Submodule.smul_le_right]
+    have inst : Finite (M ⧸ N.toAddSubgroup) := ‹_›
+    exact mul_ne_zero this AddSubgroup.index_ne_zero_of_finite
+  let e : (N ⧸ (I • N).comap N.subtype) ≃ₗ[R] (R ⧸ I) ⊗[R] N :=
+    Submodule.quotEquivOfEq _ (I • (⊤ : Submodule R N)) (Submodule.map_injective_of_injective
+      N.injective_subtype (by simp [Submodule.smul_le_right])) ≪≫ₗ
+        (quotTensorEquivQuotSMul N I).symm
+  rw [Nat.card_congr e.toEquiv]
+  have : Module.Finite R N := Module.Finite.iff_fg.mpr hN
+  have : Finite ((R ⧸ I) ⊗[R] N) := Module.finite_of_finite (R ⧸ I)
+  exact Nat.card_pos.ne'
+
+open TensorProduct in
+lemma Submodule.index_smul_le [Finite (R ⧸ I)]
+    (s : Finset M) (hs : Submodule.span R s = N) :
+    (I • N).toAddSubgroup.index ≤ I.toAddSubgroup.index ^ s.card * N.toAddSubgroup.index := by
```

Comment:
I assume you can't substitute `N` in practice?

Reply:
I'm not sure what you meant by this?

Reply:
Why isn't this
```suggestion
    (s : Finset M) :
    (I • Submodule.span R s).toAddSubgroup.index
      ≤ I.toAddSubgroup.index ^ s.card * (Submodule.span R s).toAddSubgroup.index := by
```

Reply:
Because usually you have `N` around but not the set of generators so this would be used as `apply Submodule.index_smul_le (s := <some construction>)`

Reply:
Yeah okay that's what I thought. Can you leave a comment to this effect?